Mathlib/Order/RelSeries.lean

```diff
@@ -154,6 +157,126 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
 lemma nonempty_of_infiniteDimensional [r.InfiniteDimensional] : Nonempty α :=
   ⟨RelSeries.withLength r 0 0⟩
 
+instance membership : Membership α (RelSeries r) :=
+  ⟨(· ∈ Set.range ·)⟩
+
+theorem mem_def {x : α} {s : RelSeries r} : x ∈ s ↔ x ∈ Set.range s :=
+  Iff.rfl
+
+/-- Start of a series, i.e. for `a₀ -r→ a₁ -r→ ... -r→ aₙ`, its head is `a₀`.
+
+Since a relation series is assumed to be non-empty, this is well defined. -/
+def head (x : RelSeries r) : α := x 0
+
+/-- End of a series, i.e. for `a₀ -r→ a₁ -r→ ... -r→ aₙ`, its last element is `aₙ`.
+
+Since a relation series is assumed to be non-empty, this is well defined. -/
+def last (x : RelSeries r) : α := x <| Fin.last _
+
+lemma head_mem (x : RelSeries r) : x.head ∈ x := ⟨_, rfl⟩
+
+lemma last_mem (x : RelSeries r) : x.last ∈ x := ⟨_, rfl⟩
+
+/--
+If `a₀ -r→ a₁ -r→ ... -r→ aₙ` and `b₀ -r→ b₁ -r→ ... -r→ bₘ` are two strict series
+such that `r aₙ b₀`, then there is a chain of length `n + m + 1` given by
+`a₀ -r→ a₁ -r→ ... -r→ aₙ -r→ b₀ -r→ b₁ -r→ ... -r→ bₘ`.
+-/
+@[simps]
+def append (p q : RelSeries r) (connect : r p.last q.head) : RelSeries r where
+  length := p.length + q.length + 1
+  toFun := Fin.append p q ∘ Fin.cast (by abel)
+  step i := by
+    obtain (hi|rfl|hi) :=
+      lt_trichotomy i (Fin.castLE (by linarith) (Fin.last _ : Fin (p.length + 1)))
+    · convert p.step ⟨i.1, hi⟩ <;> convert Fin.append_left p q _ <;> rfl
+    · convert connect
+      · convert Fin.append_left p q _; rfl
+      · convert Fin.append_right p q _; rfl
+    · set x := _; set y := _
+      change r (Fin.append p q x) (Fin.append p q y)
```

Comment:
Noting for later, can't you do something like
```suggestion
    · change r (Fin.append p q ?x) (Fin.append p q ?y)
```
I know this was talked about on Zulip but I don't know whether it was implemented.

Reply:
I just tried `change ... ?x`, but I don't think this works yet